rfcs/0127-issues-warnings.md

```diff
@@ -0,0 +1,306 @@
+
+## Detailed design
+[design]: #detailed-design
+
+### Package problems
+
+A new attribute is added to the `meta` section of a package: `problems`. If present, it is a an attribute set of attrsets where the keys are the problem names and the values have at least the following fields each:
+
+- `kind`: The 'kind' of the problem, see [problem kinds](#problem-kinds) for allowed values. Defaults to the attribute set key.
+- `message`: Required. A string message describing the issue with the package. The value should:
+  - Start with the "This package", "The application" or equivalent, or simply with the package name.
+  - Be capitalized (unless it starts with the package name).
+  - Use a period at the end.
+- `urls`: Optional, list of strings. Can be used to link issues, pull requests and other related items.
+
+Problem kinds may specify additional attributes. Apart from that, other attributes are not allowed.
+
+Example values:
+
+```nix
+meta.problems = {
+  # This one will have the name "python2-eol"
+  python2-eol = {
+    kind = "deprecated";
+    message = "This package depends on Python 2, which has reached end of life.";
+    urls = [ "https://github.com/NixOS/nixpkgs/issues/148779" ];
+  };
+  # This one will have the name "removal"
+  removal = {
+    # kind = "removal"; # Inferred from attribute key
+    message = "The application has been abandoned upstream, use libfoo instead.";
+  };
+};
+```
+
+### Problem kinds
+
+At the moment, the following values for the `kind` field of a problem are known:
+
+- `removal`: The package is scheduled for removal some time in the future. Only up to one instance per package.
+  - Packages are already being removed for various reasons on a regular basis. Now we can properly warn users about this in advance.
+- `deprecated`: The package has been abandoned upstream or has end of life dependencies.
+  - Currently we do mass-pings for all maintainers of affected packages, although many packages are under-maintained. This allows to warn users who depend on that package as last resort stake holders.
+- `maintainerless`: `meta.maintainers` is empty. Cannot be manually declared in `meta.problems`. Only up to one instance per package.
+  - This is currently provided by `config.showDerivationWarnings`, which is only used by Hydra.
+- `insecure`: (Reserved for future use.) The package has some security vulnerabilities.
+  - This will hopefully replace the existing insecure warnings (`meta.knownVulnerabilities`, `config.permittedInsecurePackages`) in the future (see Future work).
+- `broken`: (Reserved for future use.) The package is broken in some way.
+  - This will hopefully replace or enhance `meta.broken` in the future (see Future work).
+- `unsupported`: (Reserved for future use.) The package is not expected to build on this platform.
+  - This will hopefully replace `meta.platforms` and `meta.badPlatforms` in the future (see Future work).
```

Comment:
Should there be a reservation for unfree licenses? In other places `unfree` is listed among the things to handle…

(This is a minor nitpick and I believe that neither resolution of it should affect FCP)

Reply:
I had it in there in the past, however `unfree` is one of the most complicated ones to adapt due to its powerful licensing and predicate options. It *might* be a thing to happen in the future, and I really hope so, but honestly I can't make any promises here. Maybe I'll see a path for it after getting experience with implementing the others (which I'm a lot more optimistic about generally), we'll see.